### DISCUSSION TITLE
Using PPP as a tunneling layer

### DISCUSSION THREAD

**Discussion by ValdikSS** [shadowsocks/shadowsocks-org#244] [General]
Anti-censorship proxy software is always implemented as proxy: it has a HTTP or Socks front-end, it works only with TCP and sometimes UDP traffic, it can't route incoming connections to the client, it requires configuration of every software to use it, or to set up a pseudo-VPN with its own userspace IP stack, which does not bypass all the limitations mentioned. However, it is easy to program, as the programmer is presented with data flows, not individual packets.

VPN software provides an interface capable of encapsulating any IP protocol (TCP/UDP/ICMP/GRE/etc), you don't need to configure applications separately to use it, and the applications are able to accept incoming connections, has better compatibility with VoIP/WebRTC/NAT Punching. However it's inconvenient/non-trivial for the programmer to handle individual packets.

The question which kept my mind busy whole last month is: why the developers of anti-censorship software have never considered PPP as a building block for tunneling layer? PPP is universally implemented in modern OS, provides all the advantages of the VPN and additionally authentication, IP assignment, compression layers, and ease of programming as in proxy — the developer of anti-censorship layer don't need to implement all of these things, just send and receive PPP frames generated by the OS.

Wouldn't it be wise to use PPP as the endpoint? If not in a 'raw' form, then as L2TP — a PPP-over-UDP layer which could listen on 127.0.0.1:1701 (unprivileged port) on the local machine. This would also allow to configure not only desktop OS, but also routers with L2TP support (also univerally supported). There are lots of enterprise-grade PPP termination systems.

Would like to hear your opinion.

**Comment by marknefedov** [comment]
I’m not that well versed in networking terminology by PPP do you mean this https://en.m.wikipedia.org/wiki/Point-to-Point_Protocol ?

EDIT: I’ve originally thought about Microsoft’s PPTP

**Reply by ValdikSS** [reply]
No, Microsoft’s Point to Point Tunneling Protocol is PP**T**P (which is also based on PPP, https://en.wikipedia.org/wiki/Point-to-Point_Protocol)

Plain PPP is designed to be used over physical wire as L2 protocol and was used for dial-up networks. Later other protocols appeared to run over L2/L3 networks, such as PPTP and L2TP, which are also based on PPP.

**Reply by marknefedov** [reply]
Thank you very much! so the closest thing to what you describing is probably this: 
https://github.com/pasis/pppoat ?

**Reply by ValdikSS** [reply]
Yes, pppoat is an implementation of what I describe. However, OS provides PPP configuration by itself, which could be integrated directly into anti-censorship program.
Or the anti-censorship program may provide L2TP server (which is PPP with some additional protocol), which could be used even over the network (you can configure another PC in LAN to connect to this L2TP server, or even a router itself).

**Reply by marknefedov** [reply]
I'm still barely understand what is happening here but, I've been researching this at my spare time, and bumped on https://stackoverflow.com/questions/35998803/pppd-over-socat-using-interface-option, so essentially, all we need to make "PPP over anything" (not the project above) work, is for this "anything" is to provide a serial link (PPPoAT provides a linux PIPE, and socat provides PTY) that pppd can pipe data in/out, so pppd will do auth/sessions and other heavy lifting for us?

**Reply by ValdikSS** [reply]
Yes, yes, exactly. It's just a matter of using stdin/stdout pipe for pppd to get a real VPN connection between two proxied endpoints.

**Reply by marknefedov** [reply]
By the way, is there anything like this for l2tp? I found [xl2tp](https://github.com/xelerance/xl2tpd), but as I undertood it routes ppp over l2tp.

**Reply by marknefedov** [reply]
I wonder this approach can be transferred to smartphones. [IOS](https://developer.apple.com/documentation/networkextension/packet_tunnel_provider) at least have something resembling TAP device, but on [Android](https://issuetracker.google.com/issues/165313043) only option is TUN. 
It may be possible to create a local l2tp/ipsec server and connect to it.

**Reply by ValdikSS** [reply]
> By the way, is there anything like this for l2tp? I found [xl2tp](https://github.com/xelerance/xl2tpd), but as I undertood it routes ppp over l2tp.

You mean an L2TP-only server with support for remote PPPd tunneling? I don't know any, but there should be some for certain.

> It may be possible to create a local l2tp/ipsec server and connect to it.

It should be possible to make a plain L2TP server (it listens on non-privileged port, while for IPsec you need port 500 which is privileged, not possible on mobile without root).

**Reply by ValdikSS** [reply]
Maybe accel-l2tp supports this type of configuration.
